src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,132 @@
     public T get(int index) {
-        return null;
+        if (index < size && index >= 0) {
+            return (T) this.internalArray[index];
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("The such index doesn't exist");
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < size && index >= 0) {
+            internalArray[index] = value;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Wrong array index!");
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T oldelement = null;
+        if (index < this.size && index >= 0) {
+            oldelement = get(index);
+            Object[] newFormedArray = new Object[internalArray.length];
+            System.arraycopy(internalArray, 0, newFormedArray, 0, index);
+            System.arraycopy(internalArray, index + 1, newFormedArray, index, (size - index) - 1);
+            internalArray = newFormedArray;
+            size--;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Wrong index! Index '"
+                    + index + "' doesn't exist");
+        }
+        return oldelement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T oldElement = null;
+        for (int i = 0; i < size; i++) {
+            if (internalArray[i] != null && internalArray[i].equals(element)) {
+                oldElement = element;
+                remove(i);
+                break;
+            } else if (internalArray[i] == null && internalArray[i] == element) {
+                remove(i);
+                break;
+            }
+        }
+        if (oldElement == null && element != null) {
+            throw new NoSuchElementException("This " + element + " wasn't found");
+        }
+        return oldElement;
     }
 
     @Override
     public int size() {
-        return 0;
+        return this.size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void grow() {
+        Object[] bufferArray = new Object[size + (size >> 1)];
+        System.arraycopy(internalArray, 0, bufferArray, 0, size);
+        internalArray = bufferArray;
+    }
+
+    private void shiftToRight(int index) {
+        Object[] newBufferedArray = new Object[size + 1];
+        if ((size + 1) <= internalArray.length) {
+            System.arraycopy(internalArray, 0, newBufferedArray, 0, index + 1);
+            System.arraycopy(internalArray, index, newBufferedArray, index + 1, size - index);
+            internalArray = newBufferedArray;
+        } else if ((size + 1) > internalArray.length) {
+            grow();
+            System.arraycopy(internalArray, 0, newBufferedArray, 0, index + 1);
+            System.arraycopy(internalArray, index, newBufferedArray, index + 1, size - index);
+            internalArray = newBufferedArray;
```

Comment:
you need a maximum of one system.arraycopy call for each method - hint, you can use the same array twice, as src and dst